src/compiler/checker.ts

```diff
@@ -15014,6 +15014,9 @@ namespace ts {
                     assumeTrue = !assumeTrue;
                 }
                 const valueType = getTypeOfExpression(value);
+                if ((type.flags & TypeFlags.Unknown) && (operator === SyntaxKind.EqualsEqualsEqualsToken) && (valueType.flags & TypeFlags.Unit)) {
+                    return assumeTrue ? valueType : type;
+                }
                 if (valueType.flags & TypeFlags.Nullable) {

```

Comment:
Need to handle
```typescript
function foo2(x: unknown) {
    let y : string = 10;
    if (x === y) {
        x // x should be string here
    }
}
```

Reply:
Ah yes, thankyou. I think I can fix this by replacing the `Unit` flag with `Primitive`, although I'm not sure what to do in other cases involving interface/object literal types e.g.

```ts
function foo2(x: unknown) {
    let y: { x: string } = { x: "hello" }
    if (x === y) {
        const z = x // what should x be here? unknown, object, or { x: string }?
    }
}
```

Reply:
To answer my own question I think it should be `object`. To narrow to `{ x: string }` would be strange because it would mean you get a more precise narrowing if `x` is `unknown` rather than `object`!

Reply:
@sheetalkamat Thank you! I've fixed this now I hope.